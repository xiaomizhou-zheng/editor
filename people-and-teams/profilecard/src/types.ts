import React from 'react';

import { IntlShape } from 'react-intl-next';

import {
  AnalyticsEventPayload,
  CreateUIAnalyticsEvent,
} from '@atlaskit/analytics-next';

import TeamCentralCardClient from './client/TeamCentralCardClient';
import TeamProfileCardClient from './client/TeamProfileCardClient';
import UserProfileCardClient from './client/UserProfileCardClient';

export interface ApiClientResponse {
  User: {
    id: string;
    isBot: boolean;
    isCurrentUser: boolean;
    avatarUrl: string | null;
    email: string | null;
    fullName: string | null;
    location: string | null;
    meta: string | null;
    nickname: string | null;
    companyName: string | null;
    remoteTimeString: string | null;
    remoteWeekdayIndex: string | null;
    remoteWeekdayString: string | null;
    status: StatusType;
    statusModifiedDate: number | null;
  };
}

export interface Team {
  // Header image props
  largeAvatarImageUrl?: string;
  smallAvatarImageUrl?: string;
  largeHeaderImageUrl?: string;
  smallHeaderImageUrl?: string;
  // Regular team details
  id: string;
  displayName: string;
  description: string;
  organizationId?: string;
  members?: {
    id: string;
    fullName: string;
    avatarUrl: string;
  }[];
}

export interface ProfileCardClientData {
  isBot: boolean;
  isCurrentUser: boolean;
  avatarUrl?: string;
  email?: string;
  fullName?: string;
  location?: string;
  meta?: string;
  nickname?: string;
  companyName?: string;
  timestring?: string;
  status: StatusType;
  statusModifiedDate?: number | null;
  customLozenges?: LozengeProps[];
}

export interface ReportingLinesUserPII {
  name: string;
  picture?: string;
}

export interface ReportingLinesUser {
  accountIdentifier: string;
  identifierType: 'ATLASSIAN_ID' | 'BASE64_HASH' | 'UNKNOWN';
  pii?: ReportingLinesUserPII;
}

export interface TeamCentralReportingLinesData {
  managers?: ReportingLinesUser[];
  reports?: ReportingLinesUser[];
}

export interface ProfileCardResourcedProps {
  userId: string;
  cloudId: string;
  resourceClient: ProfileClient;
  actions?: ProfileCardAction[];
  reportingLinesProfileUrl?: string;
  onReportingLinesClick?: (user: ReportingLinesUser) => void;
  position?: ProfilecardTriggerPosition;
  trigger?: TriggerType;
  children?: React.ReactNode;
  addFlag?: (flag: any) => void;
}

export interface ProfileCardResourcedState {
  visible?: boolean;
  isLoading?: boolean;
  hasError: boolean;
  error?: ProfileCardErrorType;
  data: ProfileCardClientData | null;
  reportingLinesData?: TeamCentralReportingLinesData;
  isKudosEnabled?: boolean;
  kudosDrawerOpen: boolean;
}

export interface ProfileCardTriggerProps {
  userId: string;
  /**
    A cloudId can be provided, and we'll verify that the target userId is an
    actual user in the specified site.

    Instead you can omit the cloudId and we won't do such a check.

    If you have a cloudId and only want to show users who are in that site
    then please provide it. If you're a site-less product or don't care about
    verifying that the shown user is in a particular site, don't provide a
    cloudId.
   */
  cloudId?: string;
  resourceClient: ProfileClient;
  actions?: ProfileCardAction[];
  reportingLinesProfileUrl?: string;
  onReportingLinesClick?: (user: ReportingLinesUser) => void;
  position?: ProfilecardTriggerPosition;
  trigger?: TriggerType;
  children?: React.ReactNode;
  testId?: string;
  addFlag?: (flag: any) => void;
}

export interface ProfileCardTriggerState {
  visible?: boolean;
  isLoading?: boolean;
  hasError: boolean;
  error?: ProfileCardErrorType;
  data: ProfileCardClientData | null;
  reportingLinesData?: TeamCentralReportingLinesData;
  shouldShowGiveKudos?: boolean;
  teamCentralBaseUrl?: string;
  kudosDrawerOpen: boolean;
}

export interface TeamProfileCardTriggerState {
  visible?: boolean;
  isLoading?: boolean;
  hasError: boolean;
  error?: any;
  data: Team | null;
  renderError?: boolean;
  shouldShowGiveKudos?: boolean;
  teamCentralBaseUrl?: string;
  kudosDrawerOpen: boolean;
}

export interface TeamProfilecardCoreProps {
  /**
    The id of the user viewing the profile card.

    This is used to determine whether to say that the member count is
    "including you" or not.
   */
  viewingUserId?: string;
  /**
    A list of extra buttons to be displayed at the bottom of the card.
    View Profile is always included by default.
   */
  actions?: ProfileCardAction[];
  /**
    A function allowing products to provide an href for the user avatars in the
    profilecard, e.g. so they can link to user's profile pages.
   */
  generateUserLink?: (userId: string) => string;
  /**
    A function allowing products to provide an onClick handler for when the
    user clicks on a user's avatar or avatar group item.
   */
  onUserClick?: (userId: string, event: React.MouseEvent<Element>) => void;
  /**
    This should be a link to the team's profile page. This will be used for:

    - Wrapping the trigger in a link to the team profile page (unless
      triggerLinkType is `none`).

    - Providing the link for the View Profile action button on the card.
   */
  viewProfileLink: string;
  /**
    An onClick action that navigates to the team's profile page. Something you
    may want, e.g. for an SPA site or tracking analytics of navigation. This
    is optional, just the viewProfileLink will suffice. Will be used for:

    - Adding an onClick to the trigger if the triggerLinkType is
      `clickable-link`.

    - Providing an onClick for the View Profile action button on the card.
   */
  viewProfileOnClick?: (event?: React.MouseEvent<Element>) => void;
}

export interface TeamProfileCardTriggerProps extends TeamProfilecardCoreProps {
  /** The id of the team. */
  teamId: string;
  /**
    The id of the organization that the team belongs to.
    Currently this is unused, but will become necessary in the future.
   */
  orgId: string;
  /** An instance of ProfileClient. */
  resourceClient: ProfileClient;
  /**
    The position relative to the trigger that the card should be displayed in.
   */
  position?: ProfilecardTriggerPosition;
  /**
    The interaction method used to trigger the team profile card to appear.

    - Click is generally recommended, but your needs may vary.

    - Hover works for mouse users, but does not support those who use a
      keyboard or screen reader, avoid using this if it's possible or makes
      sense.

    - Hover-click is usable for scenarios like inline-edits, where mouse users
      cannot click on the trigger without causing side effects, but keyboard
      users are still able to navigate into and trigger the profile card.

    Look at the "Team Profilecard Trigger" or "Trigger Link Types" examples to
    see how they behave, or ask in #team-twp-people-teams on Slack for our
    recommendations.
   */
  trigger?: 'hover' | 'click' | 'hover-click';
  /**
    We generally prefer to wrap the trigger in a link to the team profile
    page. This prop determines how that link behaves.

    - Link is generally the recommended prop (especially in combination with
      click or hover-click for the trigger prop above). It wraps the trigger in
      an anchor tag with the team profile link (that users can interact with
      via middle-click, etc.), but left clicking on the link is suppressed.

    - None does not wrap the trigger in a link at all. This makes it difficult
      for keyboard or screen reader users to know how to trigger the profile
      card. Generally avoid this.

    - Clickable-link wraps the trigger in a link with no special behaviour.
      This is suitable for places where you want the trigger to serve primarily
      as a link, and optionally allow hovering to preview the team first.

    Look at the example on "Trigger Link Types" for more in-depth analysis, or
    ask in #team-twp-people-teams on Slack for our recommendations.
   */
  triggerLinkType?: 'none' | 'link' | 'clickable-link';
  /**
    This is the component that will cause a team profile card to appear when
    interacted with according to the method specified by the trigger prop.
   */
  children?: React.ReactNode;
  /**
   * Used by the card to show Flags.
   */
  addFlag?: (flag: any) => void;
  /**
   * Optional cloudId. Pass this if rendering card within a sited context.
   */
  cloudId?: string;
}

export type StatusType = 'active' | 'inactive' | 'closed';

export type TriggerType = 'hover' | 'click';

export type StatusModifiedDateType =
  | 'noDate'
  | 'thisWeek'
  | 'thisMonth'
  | 'lastMonth'
  | 'aFewMonths'
  | 'severalMonths'
  | 'moreThanAYear';

export interface ProfileCardAction {
  callback?: (...args: any[]) => any;
  shouldRender?: (data: any) => boolean;
  id?: string;
  label: React.ReactNode;
  // Link to provide general link behaviour to the button. If both link and callback are provided link behaviour will be suppressed on click.
  link?: string;
}

export type LozengeColor =
  | 'default'
  | 'success'
  | 'removed'
  | 'inprogress'
  | 'new'
  | 'moved';

export interface LozengeProps {
  text: React.ReactNode;
  appearance?: LozengeColor; // defaults to 'default'
  isBold?: boolean; // defaults to false
}

export interface ProfilecardProps {
  isLoading?: boolean;
  hasError?: boolean;
  errorType?: ProfileCardErrorType;
  status?: StatusType;
  isBot?: boolean;
  avatarUrl?: string;
  fullName?: string;
  meta?: string;
  userId?: string;
  isCurrentUser?: boolean;
  // Nick name is also known as public name
  nickname?: string;
  email?: string;
  location?: string;
  companyName?: string;
  timestring?: string;
  actions?: ProfileCardAction[];
  clientFetchProfile?: () => void;
  statusModifiedDate?: number | null;
  withoutElevation?: boolean;
  /** Show manager and direct reports section on profile hover card, if available */
  reportingLines?: TeamCentralReportingLinesData;
  /** Base URL to populate href value for manager's and direct reports' user avatar  */
  reportingLinesProfileUrl?: string;
  /** Click handler when user clicks on manager's and direct reports' user avatar, un-clickable otherwise */
  onReportingLinesClick?: (user: ReportingLinesUser) => void;
  isKudosEnabled?: boolean;
  teamCentralBaseUrl?: string;
  addFlag?: (flag: any) => void;
  cloudId?: string;

  // Allow to pass custom message for disabled account which `status` prop is `inactive` or `closed`.
  // `disabledAccountMessage` should not contain react-intl-next components, ex: `FormattedMessage`,
  // because ProfileCard component is wrapped in its own `IntlProvider` and `FormattedMessage` will loads messages of `@atlaskit/profilecard`,
  // not from the consumer of `@atlaskit/profilecard`.
  disabledAccountMessage?: React.ReactNode;
  // Allow to show a status lozenge for disabled account which `status` prop is `inactive` or `closed`
  hasDisabledAccountLozenge?: boolean;
  // Allow consumers to pass in custom lozenges that will be displayed under the heading
  customLozenges?: LozengeProps[];
  openKudosDrawer?: () => void;
}

export type AnalyticsFromDuration = (duration: number) => AnalyticsEventPayload;

export type AnalyticsFunction = (generator: AnalyticsFromDuration) => void;

export interface AnalyticsProps {
  createAnalyticsEvent?: CreateUIAnalyticsEvent;
}

export interface AnalyticsWithDurationProps {
  fireAnalyticsWithDuration: AnalyticsFunction;
}

export interface TeamProfilecardProps extends TeamProfilecardCoreProps {
  /** Indicates whether the team's details are still loading. */
  isLoading?: boolean;
  /** Indicates whether an error occurred whilst fetching team details. */
  hasError?: boolean;
  /** Describes the type of error that occurred, if any. */
  errorType?: ProfileCardErrorType;
  /** The details of the team to be shown. */
  team?: Team;
  /** A callback that will try to re-fetch data in case an error occurred. */
  clientFetchProfile?: () => void;
  /** Details relevant to passing around analytics. */
  analytics: AnalyticsFunction;
}

export interface MessageIntlProviderProps {
  children: React.ReactNode;
  intl: IntlShape;
}

export type RelativeDateKeyType =
  | 'ThisWeek'
  | 'ThisMonth'
  | 'LastMonth'
  | 'AFewMonths'
  | 'SeveralMonths'
  | 'MoreThanAYear'
  | null;

export interface ProfileClient {
  flushCache: () => void;
  getProfile: (
    cloudId: string,
    userId: string,
    analytics?: (event: AnalyticsEventPayload) => void,
  ) => Promise<ProfileCardClientData>;
  getTeamProfile: (
    teamId: string,
    orgId?: string,
    fireAnalytics?: (event: AnalyticsEventPayload) => void,
  ) => Promise<Team>;
  getReportingLines: (userId: string) => Promise<TeamCentralReportingLinesData>;
  shouldShowGiveKudos: () => Promise<boolean>;
  getTeamCentralBaseUrl: () => string | undefined;
}

export type ProfilecardTriggerPosition =
  | 'bottom-start'
  | 'bottom'
  | 'bottom-end'
  | 'left-start'
  | 'left'
  | 'left-end'
  | 'top-end'
  | 'top'
  | 'top-start'
  | 'right-end'
  | 'right'
  | 'right-start';

export type ProfileCardErrorType = {
  reason: 'default' | 'NotFound';
} | null;

export interface ProfileClientOptions {
  url: string;
  gatewayGraphqlUrl?: string;
  cacheSize?: number;
  cacheMaxAge?: number;
  /** Enables Team Central functionality if enabled e.g. /gateway/api/watermelon/graphql*/
  teamCentralUrl?: string;
  /** URL to the Team Central app e.g. team.atlassian.com */
  teamCentralBaseUrl?: string;
  /** Name of integrating product e.g. jira, atlas, confluence **/
  productIdentifier?: string;
  cloudId?: string;
}

export interface ClientOverrides {
  userClient?: UserProfileCardClient;
  teamClient?: TeamProfileCardClient;
  teamCentralClient?: TeamCentralCardClient;
}
