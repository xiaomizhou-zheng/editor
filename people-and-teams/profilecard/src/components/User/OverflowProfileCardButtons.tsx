import React, { useCallback, useState } from 'react';

import { useIntl } from 'react-intl-next';

import Button from '@atlaskit/button/custom-theme-button';
import DropdownMenu, {
  DropdownItem,
  DropdownItemGroup,
} from '@atlaskit/dropdown-menu';
import MoreIcon from '@atlaskit/icon/glyph/more';

import messages from '../../messages';
import { OverflowActionButtonsWrapper } from '../../styled/Card';
import { AnalyticsWithDurationProps, ProfileCardAction } from '../../types';
import { moreActionsClicked } from '../../util/analytics';

type OverflowButtonsProps = {
  actions: ProfileCardAction[];
  onItemClick: (
    action: ProfileCardAction,
    args: any,
    event: React.MouseEvent | React.KeyboardEvent,
    index: number,
  ) => void;
} & AnalyticsWithDurationProps;

export const ACTION_OVERFLOW_THRESHOLD = 2;

export const OverflowProfileCardButtons = (props: OverflowButtonsProps) => {
  const intl = useIntl();

  const [, setOpen] = useState(false);

  const { actions, onItemClick, fireAnalyticsWithDuration } = props;

  const numActions = actions.length + ACTION_OVERFLOW_THRESHOLD;

  const onOpenChange = useCallback(
    ({ isOpen: nextOpen }: { isOpen: boolean }) => {
      setOpen((prevOpen) => {
        if (nextOpen && !prevOpen) {
          fireAnalyticsWithDuration((duration) =>
            moreActionsClicked('user', {
              duration,
              numActions,
            }),
          );
        }

        return nextOpen;
      });
    },
    [numActions, fireAnalyticsWithDuration],
  );

  return (
    <OverflowActionButtonsWrapper data-testid="profilecard-actions-overflow">
      <DropdownMenu
        onOpenChange={onOpenChange}
        placement={'bottom-end'}
        trigger={({ triggerRef, isSelected, testId, ...providedProps }) => (
          <Button
            type="button"
            {...providedProps}
            ref={triggerRef}
            iconBefore={
              <MoreIcon
                label={intl.formatMessage(messages.profileCardMoreIconLabel)}
              />
            }
          />
        )}
      >
        <DropdownItemGroup>
          {actions.map((action, index) => (
            <DropdownItem
              key={action.id}
              onClick={(event, ...args: any) => {
                onItemClick(action, args, event, index);
              }}
              href={action.link}
            >
              {action.label}
            </DropdownItem>
          ))}
        </DropdownItemGroup>
      </DropdownMenu>
    </OverflowActionButtonsWrapper>
  );
};
