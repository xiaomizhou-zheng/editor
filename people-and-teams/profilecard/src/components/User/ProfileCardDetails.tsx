import React from 'react';

import { FormattedMessage } from 'react-intl-next';

import Lozenge from '@atlaskit/lozenge';

import relativeDate from '../../internal/relative-date';
import messages from '../../messages';
import {
  AppTitleLabel,
  CustomLozengeContainer,
  DetailsGroup,
  DisabledInfo,
  FullNameLabel,
  JobTitleLabel,
  LozengeWrapper,
} from '../../styled/Card';
import {
  AnalyticsWithDurationProps,
  LozengeProps,
  ProfilecardProps,
} from '../../types';
import { IconLabel } from '../Icon';

import ReportingLinesDetails from './ReportingLinesDetails';

const renderName = (nickname?: string, fullName?: string, meta?: string) => {
  if (!fullName && !nickname) {
    return null;
  }

  const isNicknameRedundant = !nickname || nickname === fullName;
  const shownNickname = ` (${nickname}) `;

  const displayName = isNicknameRedundant
    ? fullName
    : `${fullName}${shownNickname}`;

  return (
    <FullNameLabel noMeta={!meta} data-testid="profilecard-name">
      {displayName}
    </FullNameLabel>
  );
};

const disabledAccountDesc = (
  statusModifiedDate: number | null | undefined,
  disabledAccountMessage: React.ReactNode | undefined,
  status: 'closed' | 'inactive' = 'closed',
) => {
  // consumer does not want to use built-in message
  if (disabledAccountMessage) {
    return disabledAccountMessage;
  }

  const date = statusModifiedDate ? new Date(statusModifiedDate * 1000) : null;
  const relativeDateKey = relativeDate(date);

  const msgKey = (relativeDateKey
    ? `${status}AccountDescMsgHasDate${relativeDateKey}`
    : `${status}AccountDescMsgNoDate`) as keyof typeof messages;

  const secondSentence = <FormattedMessage {...messages[msgKey]} />;

  return (
    <p data-testid="profilecard-disabled-account">
      <FormattedMessage {...messages.generalDescMsgForDisabledUser} />{' '}
      {secondSentence}
    </p>
  );
};

const CustomLozenges = ({ lozenges = [] }: { lozenges?: LozengeProps[] }) => {
  if (lozenges.length === 0) {
    return null;
  }

  return (
    <CustomLozengeContainer>
      {lozenges.map(({ text, ...otherProps }, index) => (
        <Lozenge {...otherProps} key={index}>
          {text}
        </Lozenge>
      ))}
    </CustomLozengeContainer>
  );
};

const BotProfileCardDetails = (props: ProfilecardProps) => {
  const { fullName, nickname } = props;

  return (
    <DetailsGroup>
      {renderName(nickname, fullName)}
      <AppTitleLabel>App</AppTitleLabel>
    </DetailsGroup>
  );
};

const DisabledProfileCardDetails = (
  props: ProfilecardProps & { status: 'closed' | 'inactive' },
) => {
  const {
    companyName,
    disabledAccountMessage,
    fullName,
    hasDisabledAccountLozenge = true,
    nickname,
    status,
    statusModifiedDate,
  } = props;

  const name =
    status === 'inactive'
      ? fullName || nickname
      : nickname || (
          <FormattedMessage {...messages.disabledAccountDefaultName} />
        );

  return (
    <DetailsGroup>
      <FullNameLabel noMeta isDisabledAccount data-testid="profilecard-name">
        {name}
      </FullNameLabel>

      {hasDisabledAccountLozenge && (
        <LozengeWrapper>
          <Lozenge appearance="default" isBold>
            {status === 'inactive' ? (
              <FormattedMessage {...messages.inactiveAccountMsg} />
            ) : (
              <FormattedMessage {...messages.closedAccountMsg} />
            )}
          </Lozenge>
        </LozengeWrapper>
      )}

      <DisabledInfo>
        {disabledAccountDesc(
          statusModifiedDate,
          disabledAccountMessage,
          status,
        )}
      </DisabledInfo>

      {status === 'inactive' && (
        <IconLabel icon="companyName">{companyName}</IconLabel>
      )}
    </DetailsGroup>
  );
};

export const ProfileCardDetails = (
  props: ProfilecardProps & AnalyticsWithDurationProps,
) => {
  const { meta, status } = props;

  if (props.isBot) {
    return <BotProfileCardDetails {...props} />;
  }

  if (status === 'inactive' || status === 'closed') {
    return <DisabledProfileCardDetails {...props} status={status} />;
  }

  return (
    <DetailsGroup>
      {renderName(props.nickname, props.fullName, meta)}
      {meta && <JobTitleLabel>{meta}</JobTitleLabel>}
      <CustomLozenges lozenges={props.customLozenges} />
      <IconLabel icon="email">{props.email}</IconLabel>
      <IconLabel icon="time">{props.timestring}</IconLabel>
      <IconLabel icon="companyName">{props.companyName}</IconLabel>
      <IconLabel icon="location">{props.location}</IconLabel>
      <ReportingLinesDetails
        reportingLines={props.reportingLines}
        reportingLinesProfileUrl={props.reportingLinesProfileUrl}
        onReportingLinesClick={props.onReportingLinesClick}
        fireAnalyticsWithDuration={props.fireAnalyticsWithDuration}
      />
    </DetailsGroup>
  );
};
