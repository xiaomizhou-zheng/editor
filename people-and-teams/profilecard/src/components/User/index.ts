export { DELAY_MS_HIDE, DELAY_MS_SHOW } from '../../util/config';

export { ProfileCardLazy } from './lazyProfileCard';
import ProfileCardTrigger from './ProfileCardTrigger';

export default ProfileCardTrigger;
