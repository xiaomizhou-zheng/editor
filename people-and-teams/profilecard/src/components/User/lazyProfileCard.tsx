import { lazy } from 'react';

export const ProfileCardLazy = lazy(
  () =>
    import(
      /* webpackChunkName: "@atlaskit-internal_lazy-profilecard" */
      './ProfileCard'
    ),
);
