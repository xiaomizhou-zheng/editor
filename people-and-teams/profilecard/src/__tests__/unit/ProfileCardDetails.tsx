import React from 'react';

import { render } from '@testing-library/react';
import { shallow } from 'enzyme';
import { IntlProvider } from 'react-intl-next';

import { ProfileCardDetails } from '../../components/User/ProfileCardDetails';
import { LozengeProps, ReportingLinesUser } from '../../types';

jest.mock('react-intl-next', () => {
  return {
    ...(jest.requireActual('react-intl-next') as any),
    useIntl: jest.fn().mockReturnValue({
      locale: 'en',
      formatMessage: (descriptor: any) => descriptor.defaultMessage,
      FormattedMessage: ({ descriptor }: any) => descriptor.defaultMessage,
    }),
  };
});

type Props = Parameters<typeof ProfileCardDetails>[0];

const defaultProps: Props = {
  fireAnalyticsWithDuration: jest.fn(),
  fullName: 'full name test',
  status: 'active',
  nickname: 'jscrazy',
  companyName: 'Atlassian',
};

const renderComponent = (props: Partial<Props> = {}) =>
  render(
    <IntlProvider locale="en">
      <ProfileCardDetails {...defaultProps} {...props} />
    </IntlProvider>,
  );

describe('ProfileCardDetails', () => {
  describe('name', () => {
    describe.each([
      ['active user', false],
      ['bot account', true],
    ])('for %s', (_, isBot) => {
      it('should include nickname if applicable', () => {
        const { getByTestId } = renderComponent({
          isBot,
        });
        const component = getByTestId('profilecard-name');
        expect(component.textContent).toMatchInlineSnapshot(
          `"full name test (jscrazy) "`,
        );
      });

      it('should only show name if no nickname', () => {
        const { getByTestId } = renderComponent({
          isBot,
          nickname: undefined,
        });
        const component = getByTestId('profilecard-name');
        expect(component.textContent).toMatchInlineSnapshot(`"full name test"`);
      });

      it('should show only once if nickname and full name match', () => {
        const { getByTestId } = renderComponent({
          isBot,
          fullName: 'Same name',
          nickname: 'Same name',
        });
        const component = getByTestId('profilecard-name');
        expect(component.textContent).toMatchInlineSnapshot(`"Same name"`);
      });

      it('should show nothing if no name or nickname provided', () => {
        const { queryByTestId } = renderComponent({
          isBot,
          fullName: undefined,
          nickname: undefined,
        });
        const component = queryByTestId('profilecard-name');
        expect(component).toBeNull();
      });
    });

    describe('for inactive user', () => {
      it('should show full name if available', () => {
        const { getByTestId } = renderComponent({
          status: 'inactive',
        });
        const component = getByTestId('profilecard-name');
        expect(component.textContent).toMatchInlineSnapshot(`"full name test"`);
      });

      it('should show nickname if no full name', () => {
        const { getByTestId } = renderComponent({
          status: 'inactive',
          fullName: undefined,
        });
        const component = getByTestId('profilecard-name');
        expect(component.textContent).toMatchInlineSnapshot(`"jscrazy"`);
      });
    });

    describe('for deleted user', () => {
      it('should show nickname if available', () => {
        const { getByTestId } = renderComponent({
          status: 'closed',
        });
        const component = getByTestId('profilecard-name');
        expect(component.textContent).toMatchInlineSnapshot(`"jscrazy"`);
      });

      it('should show "Former user" if no nickname', () => {
        const { getByTestId } = renderComponent({
          status: 'closed',
          nickname: undefined,
        });
        const component = getByTestId('profilecard-name');
        expect(component.textContent).toMatchInlineSnapshot(`"Former user"`);
      });
    });
  });

  describe('Disabled account message', () => {
    describe.each(['inactive', 'closed'] as Props['status'][])(
      '%s user',
      (status: Props['status']) => {
        it('should show provided message', () => {
          const { getByText } = renderComponent({
            status,
            disabledAccountMessage: 'This account is disabled',
          });
          expect(getByText('This account is disabled')).toBeDefined();
        });

        it(`should show "just now" message for ${status} user`, () => {
          const { getByTestId } = renderComponent({
            status,
            statusModifiedDate: Date.now() / 1000,
          });

          expect(
            getByTestId('profilecard-disabled-account').textContent,
          ).toEqual(
            `You can no longer collaborate with this person. Their account was ${
              status === 'inactive' ? 'deactivated' : 'deleted'
            } this week.`,
          );
        });

        it(`should show message for ${status} user when modified date is not available`, () => {
          const { getByTestId } = renderComponent({
            status,
            statusModifiedDate: undefined,
          });

          expect(
            getByTestId('profilecard-disabled-account').textContent,
          ).toEqual(
            `You can no longer collaborate with this person. Their account has been ${
              status === 'inactive' ? 'deactivated' : 'deleted'
            }.`,
          );
        });

        it('should show lozenge if enabled', () => {
          const { getByText } = renderComponent({
            status,
            hasDisabledAccountLozenge: true,
          });

          expect(
            getByText(
              `Account ${status === 'closed' ? 'deleted' : 'deactivated'}`,
            ),
          ).toBeDefined();
        });

        it('should not show lozenge if disabled', () => {
          const { queryByText } = renderComponent({
            status,
            hasDisabledAccountLozenge: false,
          });

          expect(
            queryByText(
              `Account ${status === 'closed' ? 'deleted' : 'deactivated'}`,
            ),
          ).toBeNull();
        });
      },
    );
  });

  describe('Custom lozenges', () => {
    const customLozenges: LozengeProps[] = [
      {
        text: 'Guest',
        appearance: 'new',
        isBold: true,
      },
      {
        text: 'Cool Bean',
        appearance: 'removed',
      },
      {
        text: <div>Another Role</div>,
        appearance: 'inprogress',
        isBold: true,
      },
    ];

    const getTextOf = (elem: React.ReactElement | string | null): string => {
      if (typeof elem === 'string') {
        return elem.toString();
      }

      if (!elem) {
        return '';
      }

      return getTextOf(elem.props.children);
    };

    it('should render multiple lozenges as expected', () => {
      const { getByText } = renderComponent({
        customLozenges,
      });

      for (const { text } of customLozenges) {
        const childText = getTextOf(text as React.ReactElement | string);

        expect(getByText(childText)).not.toBeNull();
      }
    });

    it('should render only one lozenge if only one provided', () => {
      const { getByText } = renderComponent({
        customLozenges: [customLozenges[0]],
      });

      const childText = getTextOf(
        customLozenges[0].text as React.ReactElement | string,
      );

      expect(getByText(childText)).not.toBeNull();
    });
  });

  describe('reportingLines property', () => {
    const renderShallow = (props = {}) => {
      return shallow(<ProfileCardDetails {...defaultProps} {...props} />);
    };
    const exampleReportingLinesUser: ReportingLinesUser = {
      accountIdentifier: 'abcd',
      identifierType: 'ATLASSIAN_ID',
      pii: {
        name: 'name',
        picture: 'picture',
      },
    };

    it('should match snapshot when no reporting lines are specified', () => {
      expect(renderShallow({ reportingLines: undefined })).toMatchSnapshot();
      expect(renderShallow({ reportingLines: {} })).toMatchSnapshot();
      expect(
        renderShallow({ reportingLines: { managers: [], reports: [] } }),
      ).toMatchSnapshot();
    });

    it('should match snapshot when reporting lines are specified', () => {
      expect(
        renderShallow({
          reportingLines: {
            managers: [exampleReportingLinesUser],
            reports: [exampleReportingLinesUser, exampleReportingLinesUser],
          },
        }),
      ).toMatchSnapshot();
    });
  });
});
