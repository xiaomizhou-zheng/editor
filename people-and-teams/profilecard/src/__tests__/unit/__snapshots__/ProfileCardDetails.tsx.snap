// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileCardDetails reportingLines property should match snapshot when no reporting lines are specified 1`] = `
<styled.div>
  <styled.span
    data-testid="profilecard-name"
    noMeta={true}
  >
    full name test (jscrazy) 
  </styled.span>
  <CustomLozenges />
  <IconLabel
    icon="email"
  />
  <IconLabel
    icon="time"
  />
  <IconLabel
    icon="companyName"
  >
    Atlassian
  </IconLabel>
  <IconLabel
    icon="location"
  />
  <ReportingLinesDetails
    fireAnalyticsWithDuration={[MockFunction]}
  />
</styled.div>
`;

exports[`ProfileCardDetails reportingLines property should match snapshot when no reporting lines are specified 2`] = `
<styled.div>
  <styled.span
    data-testid="profilecard-name"
    noMeta={true}
  >
    full name test (jscrazy) 
  </styled.span>
  <CustomLozenges />
  <IconLabel
    icon="email"
  />
  <IconLabel
    icon="time"
  />
  <IconLabel
    icon="companyName"
  >
    Atlassian
  </IconLabel>
  <IconLabel
    icon="location"
  />
  <ReportingLinesDetails
    fireAnalyticsWithDuration={[MockFunction]}
    reportingLines={Object {}}
  />
</styled.div>
`;

exports[`ProfileCardDetails reportingLines property should match snapshot when no reporting lines are specified 3`] = `
<styled.div>
  <styled.span
    data-testid="profilecard-name"
    noMeta={true}
  >
    full name test (jscrazy) 
  </styled.span>
  <CustomLozenges />
  <IconLabel
    icon="email"
  />
  <IconLabel
    icon="time"
  />
  <IconLabel
    icon="companyName"
  >
    Atlassian
  </IconLabel>
  <IconLabel
    icon="location"
  />
  <ReportingLinesDetails
    fireAnalyticsWithDuration={[MockFunction]}
    reportingLines={
      Object {
        "managers": Array [],
        "reports": Array [],
      }
    }
  />
</styled.div>
`;

exports[`ProfileCardDetails reportingLines property should match snapshot when reporting lines are specified 1`] = `
<styled.div>
  <styled.span
    data-testid="profilecard-name"
    noMeta={true}
  >
    full name test (jscrazy) 
  </styled.span>
  <CustomLozenges />
  <IconLabel
    icon="email"
  />
  <IconLabel
    icon="time"
  />
  <IconLabel
    icon="companyName"
  >
    Atlassian
  </IconLabel>
  <IconLabel
    icon="location"
  />
  <ReportingLinesDetails
    fireAnalyticsWithDuration={[MockFunction]}
    reportingLines={
      Object {
        "managers": Array [
          Object {
            "accountIdentifier": "abcd",
            "identifierType": "ATLASSIAN_ID",
            "pii": Object {
              "name": "name",
              "picture": "picture",
            },
          },
        ],
        "reports": Array [
          Object {
            "accountIdentifier": "abcd",
            "identifierType": "ATLASSIAN_ID",
            "pii": Object {
              "name": "name",
              "picture": "picture",
            },
          },
          Object {
            "accountIdentifier": "abcd",
            "identifierType": "ATLASSIAN_ID",
            "pii": Object {
              "name": "name",
              "picture": "picture",
            },
          },
        ],
      }
    }
  />
</styled.div>
`;
